tecton_client/exceptions.py

```diff
@@ -1,158 +1,73 @@
-        allowed_types (Optional[tuple]): Tuple of data types allowed for the value in the map.
-        key (Optional[str]): The key provided.
-        value (Optional[str]): The value provided.
-
-    Returns:
-        str: The error message string.
-
-    """
-    if key:
-        return f"{map_type} keys can only be of (str) type. Given key for {map_type} is {key} of type {type(key)}"
-    if value:
-        allowed_types = tuple([str(val_type).split("'")[1] for val_type in allowed_types])
-        return (
-            f"{map_type} values can only be of types {allowed_types}. "
-            f"Given value for {map_type} is {value} of type {type(value)}"
-        )
-
-
-class BadRequestError(TectonServerException):
+class BadRequestError(TectonHttpException):
     """Raised when the Tecton API returns a 400 Bad Request error."""
 
     STATUS_CODE = 400
 
 
-class UnauthorizedError(TectonServerException):
+class UnauthorizedError(TectonHttpException):
     """Raised when the Tecton API returns a 401 Unauthorized error."""
 
     STATUS_CODE = 401
 
 
-class ForbiddenError(TectonServerException):
+class ForbiddenError(TectonHttpException):
     """Raised when the Tecton API returns a 403 Forbidden error."""
 
     STATUS_CODE = 403
 
 
-class NotFoundError(TectonServerException):
+class NotFoundError(TectonHttpException):
     """Raised when the Tecton API returns a 404 Not Found error."""
 
     STATUS_CODE = 404
 
 
-class ResourceExhaustedError(TectonServerException):
+class ResourceExhaustedError(TectonHttpException):
     """Raised when the Tecton API returns a 429 Resources Exhausted error."""
 
     STATUS_CODE = 429
 
 
-class ServiceUnavailableError(TectonServerException):
+class ServiceUnavailableError(TectonHttpException):
     """Raised when the Tecton API returns a 503 Service Unavailable error."""
 
     STATUS_CODE = 503
 
 
-class GatewayTimeoutError(TectonServerException):
+class GatewayTimeoutError(TectonHttpException):
     """Raised when the Tecton API returns a 504 Gateway Timeout error."""
 
     STATUS_CODE = 504
 
 
-SERVER_ERRORS: dict = {error.STATUS_CODE: error for error in TectonServerException.__subclasses__()}
+_HTTP_ERRORS: dict = {error.STATUS_CODE: error for error in TectonHttpException.__subclasses__()}
+
+
+def convert_exception(httpx_exception: httpx.HTTPStatusError) -> TectonHttpException:
+    """Convert a httpx.HTTPStatusError into a TectonHttpException for better message formatting"""
+    status_code = httpx_exception.response.status_code
+    exception_class = _HTTP_ERRORS.get(status_code)
+    if exception_class:
+        message = json.loads(httpx_exception.response.content).get("message")
+        return NotFoundError(
+            status_code=status_code, reason_phrase=httpx_exception.response.reason_phrase, message=message
+        )
+    else:
+        return TectonHttpException(status_code, httpx_exception.response.reason_phrase, "")
```

Comment:
Can you give an example that shows the before/after, trying to get a feel for what value we are actually getting here wrapping this?

Reply:
In a comment or just below? Here is an example of raising a 400 error from a bad argument. You can see both the raw error and the wrapper error. The wrapper error extracts the 'reason_phrase', which is where the bulk of the useful information is, which is a large benefit.

The different granular types of errors is a bit... extra, but the code for making them was already here from the intern, and it provides some value so seems OK to keep.

```
(test_venv) solbrigm@solbrigm- tecton-http-client-python % python examples/example_sync.py 
Traceback (most recent call last):
  File "/Users/solbrigm/repos/tecton-http-client-python/tecton_client/_internal/tecton_client.py", line 75, in get_features
    resp.raise_for_status()
  File "/Users/solbrigm/repos/tecton-http-client-python/test_venv/lib/python3.9/site-packages/httpx/_models.py", line 761, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://explore.tecton.ai/api/v1/feature-service/get-features'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/solbrigm/repos/tecton-http-client-python/examples/example_sync.py", line 12, in <module>
    resp = client.get_features(
  File "/Users/solbrigm/repos/tecton-http-client-python/tecton_client/_internal/tecton_client.py", line 77, in get_features
    raise convert_exception(exc) from exc
tecton_client.exceptions.NotFoundError: 400 Bad Request: Unknown join key: user_i, expected one of: [user_id]
```

Reply:
🤦 I blame the intern... will fix.

Reply:
```
tecton_client/_internal/tecton_client.py:80: in get_features
    resp.raise_for_status()
test_venv/lib/python3.9/site-packages/httpx/_models.py:761: in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
E   httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://explore.tecton.ai/api/v1/feature-service/get-features'
E   For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

The above exception was the direct cause of the following exception:
examples/example_sync.py:12: in <module>
    resp = client.get_features(
tecton_client/_internal/tecton_client.py:82: in get_features
    raise convert_exception(exc) from exc
E   tecton_client.exceptions.BadRequestError: 400 Bad Request: Unknown join key: user_i, expected one of: [user_id]
```